src/UuidType.php

```diff
@@ -0,0 +1,51 @@
+<?php declare(strict_types=1);
+
+namespace Fortuneglobe\Types;
+
+use Fortuneglobe\Types\Exceptions\InvalidUuidValueException;
+use Fortuneglobe\Types\Interfaces\RepresentsUuidValue;
+
+/**
+ * Class UuidType
+ * @package Fortuneglobe\Types
+ */
+abstract class UuidType extends StringType implements RepresentsUuidValue
+{
+	private const NIL          = '00000000-0000-0000-0000-000000000000';
+
+	private const UUID_PATTERN = '^[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}$';
+
+	public function __construct( string $uuid )
+	{
+		$this->guardUuidIsValid( $uuid );
+
+		parent::__construct( $uuid );
+	}
+
+	public static function generate() : RepresentsUuidValue
+	{
+		$data = random_bytes( 16 );
+
+		$data[6] = chr( ord( $data[6] ) & 0x0f | 0x40 );
+		$data[8] = chr( ord( $data[8] ) & 0x3f | 0x80 );
+
+		$uuid = vsprintf( '%s%s-%s-%s-%s-%s%s%s', str_split( bin2hex( $data ), 4 ) );
+
+		return new static( $uuid );
```

Comment:
Since the `Identifier` package I'm struggeling with the guarding in the constructor. This method is tested to generate a valid UUID but is still guarded in the constructor. Nonetheless I cannot provide a workaround here.

Reply:
Since the constructor (regardless if called by `fromString()` or userland code) is the method to create the value object, it's the right place for guarding the input value. Constructing a value object with an invalid value leads to unexpected behaviour or validation to be forgotten.

In this case it also prevents you to have faulty code in the `generate()` method.

Reply:
> In this case it also prevents you to have faulty code in the generate() method.

Yes, I know. But I also mentioned it's already tested with a unit test.